reports/spec1.tex

```diff
@@ -2,4 +2,55 @@
+Notice that, while filter is not one of the built-in operators of the \emph{pyrsistent} library, it is relatively simple to define a filter operator, such that the set returns contains exactly all of the elements of $s$, for which the Boolean predicate $p$ holds true.
+In \tlap{}, however, filtering is a language primitive, so if we can translate a python set $s$ to a \tlap{} set $\hat{s}$, and a Python predicate $p$ (of the above type) to a \tlap{} predicate $\hat{p}$, we can translate $\mathsf{pset\_filter}(p, s)$ to $\{ x \in \hat{s}\colon \hat{p}(x) \}$.
+
+We take this idea, and apply it to every definition in the file \emph{pythonic\_code\_generic.py}, attempting to identify \tlap{}-equivalents (w.r.t. semantics) for each defined function. 
+Later on, in milestone 3, we give a formal characterization of all of these equivalencies, in the form of rewriting rules, although for the purposes of this first specification, the entirety of this translation is manual.
+
+Here is an example of a top-level operator, $\mathsf{has\_block\_hash}$\footnote{Python definition available here: \url{https://github.com/saltiniroberto/ssf/blob/ad3ba2c21bc1cd554a870a6e0e4d87040558e129/high_level/common/helpers.py\#L20-L24}}, as it is expressed in \tlap{}:
+\begin{lstlisting}[language=tla,columns=fullflexible]
+has_block_hash(block_hash, node_state) ==
+    block_hash \in DOMAIN node_state.view_blocks
+\end{lstlisting}
+%
+We can see that the original Python implementation calls $\mathsf{pmap\_has}$, which checks whether a key is contained in a given map. In our translation, the map is modeled with a \tlap{} function, which has an associated $\DOMAIN$ set, so we simply check for set membership with the built-in $\backslash\mathsf{in}$ operator.
+
+Although it is relevant in the process of translation, we do not give explicit rules for the translation of Python language primitives to \tlap{} in general, since attempting to establish those for the full language would vastly exceed the scope of this project. 
+Some of these constructs are particularly relevant, and we briefly mention them here:
+\paragraph{Assignments and local variables.} There are certain idiosyncrasies to do with the fact that one language is executable, and the other is not. For instance, Python allows for arbitrary variable assignment and reassignment, as well as the introduction of local variables. There are two constructs available in \tlap{} which can be used to express variable assignment:
+\begin{itemize}
+  \item state-variable update $a' = e$
+  \item LET-IN local operator definition $\mathrm{LET}\; v \defeq e \;\mathrm{IN}\; f$
+\end{itemize}
+to avoid going in to too much detail, it is up to the translator to evaluate which of the two better captures the semantics of the Python code (with, in general, a preference for the LET-IN variant).
+\paragraph{Runtime exceptions.} Python code may throw at any time, sometimes in the form of $\mathrm{Require}$-guards within function definitions. 
```

Comment:
afaict, `$\mathrm{Require}$` has no effect overall

```suggestion
\paragraph{Runtime exceptions.} Python code may throw at any time, sometimes in the form of $\mathrm{Require}$-guards within function definitions. 
```

Reply:
Fixed in https://github.com/freespek/ssf-mc/pull/65/commits/60d86ae9ca0292a4b6cc58d3a034b49ce3dfe102